dcpy/extract/ingest.py

```diff
@@ -1,30 +1,87 @@
-from . import PARQUET_PATH, config
+from datetime import datetime
+import importlib
+import os
+from pathlib import Path
+import requests
+from urllib.parse import urlparse
 
+from dcpy.utils.logging import logger
+from dcpy.connectors.edm import recipes, publishing
+from . import TMP_DIR, PARQUET_PATH, metadata, utils
 
-# doesn't strictly need to be recipe config, but at least config-like object
-def archive_raw_dataset(import_config: config.Import):
-    """Given dataset and potentially a semantic version, ingest it from its source and archive it in its
-    raw format, with metadata, in edm-recipes/raw_data"""
-    ## logic to grab dataset based on ImportDefinition, dump locally
 
-    ## logic to "compute" recipes.Config object, similar to Config.compute in library
-    ## should be limited to metadata gathering - version, etc
-    ## leaving here as opposed to helper function because not sure if it will only depend on import definition
-    ##   or read any data from downloaded dataset
+def _download_file_web(url: str, file_name: str) -> Path:
+    """Simple wrapper to download a file using requests.get.
+    Returns filepath."""
+    TMP_DIR.mkdir(exist_ok=True)
+    file = TMP_DIR / file_name
+    logger.info(f"downloading {url} to {file}")
+    response = requests.get(url)
+    response.raise_for_status()
+    with open(file, "wb") as f:
+        f.write(response.content)
+    return file
 
-    ## logic to push local dataset and "computed" config to raw archive in s3
 
-    ## return Config, from which s3path should be computed
+def download_file_from_source(template: metadata.Template, version: str) -> Path:
+    """From parsed config template and version, download raw data from source
+    and return path of saved local file."""
+    match template.source:
+        ## Non reqeust-based methods
+        case recipes.ExtractConfig.Source.LocalFile() as local_file:
```

Comment:
curious, do we compare instances or classes here? For example, it looks like we instantiating `recipes....LocalFile()`. Also, does `local_file` refer to `template.source` or  `recipes....LocalFile()`? Don't understand

Reply:
It compares the type. I agree that the syntax is a little funky, but that's python for you. `local_file` is `template.source`, but the typer knows that it is a LocalFile (with a path)

Reply:
So the distinction being `template.source` is of type `LocalFile | Socrata | ...`, so any attributes of `LocalFile` cannot be accessed. But once we match it, `local_file` is essentially the equivalent of `cast(LocalFile(), template.source)` so it's type-safe to access it's class attributes as a `LocalFile`